# Neural network with Keras

# Part A: Building the base line model

## A1. Preparing the dataset

In [131]:
# Downloading the data and reading it into a pandas dataframe
import pandas as pd
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()


Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [132]:
# Checking how many data point we have
concrete_data.shape

(1030, 9)

In [133]:
# Before building the model, I'm going to check if there ismissing value in the dataset
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [134]:
# The data seem correct. Let's describe it now
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [135]:
# Now I'm going to split the dataset into predictors and target column
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column


In [136]:
# Quick check of predictors
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [137]:
# Saving the predictor to cols beacause I need this number to build the nnetwork
cols = predictors_norm.shape[1] # number of predictors

In [138]:
# Quick check of target
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [141]:
# Importing keras now
import keras

In [142]:
# Importing the different package
from keras.models import Sequential
from keras.layers import Dense


## A 2. Building the neural network by defining the Keras regression model

In [143]:
# define regression model wtih One hidden layer of 10 nodes, and a ReLU activation function
def regression1_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(cols,)))
    model.add(Dense(10, activation='relu'))
    
    model.add(Dense(1))
    
    # compile model using the adam optimizer and the mean squared error as the loss function
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model


## A 3. Training the model on the training data

In [144]:
# Calling the function to build the model
model1 = regression1_model()


In [145]:
# Training the model using the fit method and leaving out 30% of the dataset for validation and also 50 epochs on training datasets
model1.fit(predictors, target, validation_split=0.3, epochs=50, verbose=2)


Train on 721 samples, validate on 309 samples
Epoch 1/50
 - 1s - loss: 40148.6657 - val_loss: 11770.0300
Epoch 2/50
 - 0s - loss: 4401.1371 - val_loss: 2127.6767
Epoch 3/50
 - 0s - loss: 2472.8139 - val_loss: 1984.5700
Epoch 4/50
 - 0s - loss: 2172.2704 - val_loss: 2050.4353
Epoch 5/50
 - 0s - loss: 1993.1145 - val_loss: 1802.9500
Epoch 6/50
 - 0s - loss: 1824.1128 - val_loss: 1617.1356
Epoch 7/50
 - 0s - loss: 1673.4367 - val_loss: 1492.8139
Epoch 8/50
 - 0s - loss: 1535.7038 - val_loss: 1351.9150
Epoch 9/50
 - 0s - loss: 1411.7196 - val_loss: 1274.3200
Epoch 10/50
 - 0s - loss: 1302.9251 - val_loss: 1142.1013
Epoch 11/50
 - 0s - loss: 1208.1819 - val_loss: 1038.1482
Epoch 12/50
 - 0s - loss: 1127.8926 - val_loss: 993.1644
Epoch 13/50
 - 0s - loss: 1053.0725 - val_loss: 899.2410
Epoch 14/50
 - 0s - loss: 993.8819 - val_loss: 850.6037
Epoch 15/50
 - 0s - loss: 937.4904 - val_loss: 781.3544
Epoch 16/50
 - 0s - loss: 881.8358 - val_loss: 732.8018
Epoch 17/50
 - 0s - loss: 832.8363 - val_

## A 4. Evaluate the model on the test data

In [146]:
predictions1 = model1.predict(predictors)
# Computig the mean squared error between the predicted concrete strength and the actual concrete strength
from sklearn.metrics import mean_squared_error
y_pred=predictions1
y_true=target

mean_squared_error (y_pred,y_true)

259.5098343341006

## A 5. Creating a list of 50 mean squared error

# Part B: Normalize the data

In [147]:
# Normalizing data
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()


Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [148]:
# Saving the predictor to n_cols beacause I need this number to build the nnetwork
n_cols = predictors_norm.shape[1] # number of predictors

In [149]:
# Importing keras now
import keras
# Importing the different package
from keras.models import Sequential
from keras.layers import Dense


## B 1. Building the neural network by defining the Keras regression model

In [150]:
# define regression model wtih One hidden layer of 10 nodes, and a ReLU activation function
def regression1_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    
    model.add(Dense(1))
    
    # compile model using the adam optimizer and the mean squared error as the loss function
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model


## B 2. Training the model on the training data

In [151]:
# Calling the function to build the model
model1 = regression1_model()

In [152]:
# Training the model using the fit method and leaving out 30% of the dataset for validation and also 50 epochs on training datasets
model1.fit(predictors_norm, target, validation_split=0.3, epochs=50, verbose=2)


Train on 721 samples, validate on 309 samples
Epoch 1/50
 - 2s - loss: 1662.2882 - val_loss: 1199.0507
Epoch 2/50
 - 0s - loss: 1645.0367 - val_loss: 1186.7778
Epoch 3/50
 - 1s - loss: 1627.5184 - val_loss: 1174.3066
Epoch 4/50
 - 0s - loss: 1609.4621 - val_loss: 1161.8189
Epoch 5/50
 - 0s - loss: 1591.0016 - val_loss: 1149.1676
Epoch 6/50
 - 0s - loss: 1571.9287 - val_loss: 1136.3808
Epoch 7/50
 - 0s - loss: 1552.0031 - val_loss: 1123.1122
Epoch 8/50
 - 0s - loss: 1530.7938 - val_loss: 1109.3859
Epoch 9/50
 - 0s - loss: 1508.6930 - val_loss: 1095.3900
Epoch 10/50
 - 0s - loss: 1485.0632 - val_loss: 1080.7882
Epoch 11/50
 - 0s - loss: 1460.0976 - val_loss: 1065.4436
Epoch 12/50
 - 0s - loss: 1433.7478 - val_loss: 1048.8686
Epoch 13/50
 - 0s - loss: 1404.6933 - val_loss: 1032.4264
Epoch 14/50
 - 0s - loss: 1374.5190 - val_loss: 1014.9089
Epoch 15/50
 - 0s - loss: 1341.3076 - val_loss: 997.0330
Epoch 16/50
 - 0s - loss: 1307.0171 - val_loss: 978.5180
Epoch 17/50
 - 0s - loss: 1271.4455 -

## B 3. Evaluate the model on the test data

In [153]:
predictions1 = model1.predict(predictors_norm)
# Computig the mean squared error between the predicted concrete strength and the actual concrete strength
from sklearn.metrics import mean_squared_error
y_pred1=predictions1
y_true=target

mean_squared_error (y_pred1,y_true)

337.80959744074875

## B 4. Comparing the two mean_squared_error

The mean_squared_error of the part A (259.5098343341006) <The mean_squared_error of the part B (337.80959744074875)

# Part C Increasing the number

## C 1. Increasing the number of epochs and reapeting the part B

In [154]:
# Training the model using the fit method and leaving out 30% of the dataset for validation and also 100 epochs on training datasets
model1.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)


Train on 721 samples, validate on 309 samples
Epoch 1/100
 - 0s - loss: 314.6067 - val_loss: 370.7129
Epoch 2/100
 - 0s - loss: 305.1132 - val_loss: 360.1188
Epoch 3/100
 - 0s - loss: 296.2191 - val_loss: 350.1497
Epoch 4/100
 - 0s - loss: 287.8917 - val_loss: 340.8011
Epoch 5/100
 - 0s - loss: 280.4279 - val_loss: 331.3149
Epoch 6/100
 - 0s - loss: 273.3072 - val_loss: 322.8282
Epoch 7/100
 - 0s - loss: 266.9564 - val_loss: 314.4095
Epoch 8/100
 - 0s - loss: 260.8852 - val_loss: 306.6153
Epoch 9/100
 - 0s - loss: 255.2155 - val_loss: 299.3749
Epoch 10/100
 - 0s - loss: 249.9875 - val_loss: 292.4062
Epoch 11/100
 - 0s - loss: 245.3501 - val_loss: 285.3373
Epoch 12/100
 - 0s - loss: 240.7691 - val_loss: 279.3780
Epoch 13/100
 - 0s - loss: 236.7854 - val_loss: 273.4832
Epoch 14/100
 - 0s - loss: 232.9272 - val_loss: 268.1343
Epoch 15/100
 - 0s - loss: 229.2314 - val_loss: 262.7833
Epoch 16/100
 - 0s - loss: 225.9349 - val_loss: 257.8992
Epoch 17/100
 - 0s - loss: 222.7912 - val_loss: 253

## C2. Evaluating the model on the test data

In [155]:
predictions1 = model1.predict(predictors_norm)
# Computig the mean squared error between the predicted concrete strength and the actual concrete strength
from sklearn.metrics import mean_squared_error
y_pred1=predictions1
y_true=target

mean_squared_error (y_pred1,y_true)

113.94045342533624

## C 3. Comparing the two mean_squared_error of step B and C

The mean_squared_error of the part C (113.94045342533624) < The mean_squared_error of the part B (337.80959744074875)

# Part D: Increasing the number of hidden layers of part B

In [157]:
# define regression model wtih three hidden layer of 10 nodes, and a ReLU activation function
def regression3_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model using the adam optimizer and the mean squared error as the loss function
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model


## D 1. Training the model 

In [158]:
# Calling the function to build the model
model3 = regression3_model()

In [159]:
# Training the model using the fit method and leaving out 30% of the dataset for validation and also 50 epochs on training datasets
model3.fit(predictors_norm, target, validation_split=0.3, epochs=50, verbose=2)


Train on 721 samples, validate on 309 samples
Epoch 1/50
 - 3s - loss: 1691.3657 - val_loss: 1213.6506
Epoch 2/50
 - 0s - loss: 1667.0033 - val_loss: 1185.1137
Epoch 3/50
 - 1s - loss: 1625.8191 - val_loss: 1137.2919
Epoch 4/50
 - 0s - loss: 1550.8214 - val_loss: 1055.2910
Epoch 5/50
 - 0s - loss: 1415.1985 - val_loss: 927.6340
Epoch 6/50
 - 0s - loss: 1208.3346 - val_loss: 743.4793
Epoch 7/50
 - 0s - loss: 913.1059 - val_loss: 526.4370
Epoch 8/50
 - 0s - loss: 591.3101 - val_loss: 338.2409
Epoch 9/50
 - 1s - loss: 361.3152 - val_loss: 258.9213
Epoch 10/50
 - 0s - loss: 281.4370 - val_loss: 241.7340
Epoch 11/50
 - 0s - loss: 255.2485 - val_loss: 228.8795
Epoch 12/50
 - 0s - loss: 237.7635 - val_loss: 219.5048
Epoch 13/50
 - 0s - loss: 226.4484 - val_loss: 214.4971
Epoch 14/50
 - 0s - loss: 217.3005 - val_loss: 209.3553
Epoch 15/50
 - 0s - loss: 211.4539 - val_loss: 205.7904
Epoch 16/50
 - 0s - loss: 205.7124 - val_loss: 206.3940
Epoch 17/50
 - 0s - loss: 200.8811 - val_loss: 202.2318
E

## D 2. Evaluate the model on the test data

In [160]:
predictions3 = model3.predict(predictors_norm)
# Computig the mean squared error between the predicted concrete strength and the actual concrete strength
from sklearn.metrics import mean_squared_error
y_pred=predictions3
y_true=target

mean_squared_error (y_pred,y_true)

145.7007953279716

## D 3. Comparing the two mean_squared_error of step B and D

The mean_squared_error of the part D (145.7007953279716) < The mean_squared_error of the part B (337.80959744074875)